### ISSUE TITLE
"Git LFS objects not staged for commit", What to do with it?

### ISSUE THREAD

**Issue by bitinn** [git-lfs/git-lfs#3381]
I was just switching from a feature branch back to master, and try to git pull some remote changes, then I got this:

```
$ git pull
Updating 69578c7..fc995f4
error: Your local changes to the following files would be overwritten by merge:
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Back.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Down.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Front.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Left.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Right.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Up.png
Please commit your changes or stash them before you merge.
Aborting
```

First of all, I really don't know how I got into these *changes* because they are static files from third-party package, which I haven't modified.

So my first instinct is to discard these changes. But I can't:

```
$ git checkout -- lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Back.png
Encountered 1 file(s) that should have been pointers, but weren't:
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Back.png
```

Alright, how about `git reset`? Nope:

```
$ git reset --hard HEAD
HEAD is now at 69578c7 Merge pull request #7 from bitinn/performance-baseline
Encountered 6 file(s) that should have been pointers, but weren't:
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Back.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Down.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Front.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Left.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Right.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Up.png
```

What's being held up?

```
$ git lfs status
On branch master
Git LFS objects to be pushed to origin/master:


Git LFS objects to be committed:


Git LFS objects not staged for commit:

	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Back.png (Git: 389cafe -> File: 389cafe)
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Down.png (Git: d257966 -> File: d257966)
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Front.png (Git: 0f3face -> File: 0f3face)
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Left.png (Git: cd94cd3 -> File: cd94cd3)
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Right.png (Git: 16380e3 -> File: 16380e3)
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Up.png (Git: b6305ea -> File: b6305ea)
```

Google search results didn't give me much insight when querying: "Git LFS objects not staged for commit".

<img width="809" alt="screen shot 2018-11-14 at 15 44 17" src="https://user-images.githubusercontent.com/1484279/48467566-43471a00-e824-11e8-841a-5c9827b216cc.png">

How about searching "Encountered X file(s) that should have been pointers, but weren't"?

<img width="797" alt="screen shot 2018-11-14 at 15 47 59" src="https://user-images.githubusercontent.com/1484279/48467748-c8323380-e824-11e8-8d51-7626bcb70612.png">

<img width="788" alt="screen shot 2018-11-14 at 15 47 28" src="https://user-images.githubusercontent.com/1484279/48467754-ce281480-e824-11e8-9eac-1bc76fbd5cdb.png">

OK, that's do this `git lfs uninstall` dance:

```
$ git lfs uninstall
Hooks for this repository have been removed.
Global Git LFS configuration has been removed.

$ git reset --hard
HEAD is now at 69578c7 Merge pull request #7 from bitinn/performance-baseline

$ git lfs install
Updated git hooks.
Git LFS initialized.

$ git lfs pull
[no output]

$ git pull
Updating 69578c7..fc995f4
Fast-forward
 .gitignore                                         |     3 +-
 .vscode/settings.json                              |    43 +
....
```

TL;DR: `git lfs uninstall` then `git reset --hard` were what worked for me.

**QUESTIONS**

- I believe `git lfs migrate` could have been used to solve this? I couldn't find a decent example via google: where you don't want to push the local change, but to discard them and accept remote changes.

- So I ended up going through the `git lfs uninstall` route, is this the best way? Because it seems counter-intuitive to me.

- `git lfs status` command could have offered some suggestions on this problem?

- This flow breaks the basic assumption that `git reset --hard HEAD` always discard your changes and `git stash` always stash your changes. Neither of them helped in my case, Git LFS insists on handling this but didn't offer any suggestions, couldn't this be done better?

**Comment by bk2204** [comment]
Hey, thanks for writing in.

So here's a little bit of information about what's going on:
- When you run `git reset --hard` or any other command that performs a checkout of any sort, git runs all of the filters involved, which means it runs Git LFS.  That causes your files to be marked dirty because it turns the files which were checked in as non-pointers to be turned into pointers.  There isn't a good way for Git LFS to know that you wanted to do a `git reset --hard` or `git stash` (as opposed to something else) and just blow away your changes.
- `git lfs migrate import --skip-fetch --yes HEAD` does indeed fix this problem, because it rewrites the working tree and branches to be consistent.
- `git lfs uninstall && git reset --hard HEAD` works because it removes the hooks and filters, so you can then blow away all your changes.
- `git add -u && git commit -m "delete me" && git reset --hard SOMEREVISION` will also work.
- Essentially, as long as Git LFS is installed, it will try very hard to keep files in the proper state, and if your working tree is not in the correct state, it will make it dirty in order to put it in the right state.

I'm not exactly sure what output we could produce in `git lfs status` that would be helpful here, since it doesn't have any knowledge about what's going on.  Could you say a few words about what you'd expect it to say here?

**Comment by bitinn** [comment]
Good answers, thx!

> Could you say a few words about what you'd expect it to say here?

`git lfs status` could have predicted what users might want to do, like, if there are *Git LFS objects not staged for commit*, it could add: "To discard local changes and overwrite with remote changes, run `git lfs migrate import --skip-fetch --yes HEAD`"

> `git lfs migrate import --skip-fetch --yes HEAD` does indeed fix this problem, because it rewrites the working tree and branches to be consistent.

My problem with current `git lfs status` is that there are no ways to figure out this solution exists and works like it should. I have read the manual beforehand, it doesn't offer an example where you want to discard local changes, and no usage example is offered for `--skip-fetch`.

https://github.com/git-lfs/git-lfs/blob/master/docs/man/git-lfs-migrate.1.ronn

> `git add -u && git commit -m "delete me" && git reset --hard SOMEREVISION` will also work.

It's a bit counter-intuitive as people don't usually commit something just to throw them away, `git stash` is much more intuitive but didn't work in this case.

> `git lfs uninstall && git reset --hard HEAD` works because it removes the hooks and filters, so you can then blow away all your changes.

In short, I wanted to do what you described, but I wasn't able to figure out how `git lfs migrate` could be used to achieve that, and my Google result lands me into `git lfs uninstall` flow instead of the preferred `git lfs migrate` solution.

I think `git lfs status` could be used to advertise such solution exists.

**Comment by bitinn** [comment]
Another hindsight: **when your repo is not working like it should, you usually don't want to do anything you don't understand.**

So while I did try `git lfs migrate import`, I didn't experiment with `--skip-fetch`, hence I couldn't discover this command's usage.

(And I have been using git lfs for over 2 years.)

**Comment by bk2204** [comment]
After thinking about this for a while, the best way to handle this might be a `git lfs disable` command where one could `git lfs disable reset --hard` and temporarily disable Git LFS in order to run `git reset --hard`.  That might be the most intuitive way forward.

@git-lfs/core What do you think?

**Comment by PastelMobileSuit** [comment]
@bk2204 Hmm, I'm not sure how I feel about that - from a UX perspective, `git lfs disable reset --hard` feels a bit long and unintuitive.

Saying that, I'm not sure what a good alternative would be. We could add a `git lfs reset` command but that seems worse, as it comes with the baggage of having to implement all of `git reset`.

Maybe something like `git lfs restore-pointers <rev>`? That can take any dirty files that should be pointers and convert them back.

No matter what, I think it would be worthwhile to look into adding some info on this and how to fix it to an error message somewhere. Perhaps something like the following:

```
$ git reset --hard HEAD
HEAD is now at 69578c7 Merge pull request #7 from bitinn/performance-baseline
Encountered 6 file(s) that should have been pointers, but weren't:
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Back.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Down.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Front.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Left.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Right.png
	lwrp-performance2/Assets/Leocities/Textures/Sky/01A_Day_Sunless_Up.png

To restore these files, run `git lfs migrate import --skip-fetch --yes HEAD`
```

**Comment by bitinn** [comment]
Some comments:

- IMO: unless the new commands are as intuitive as `git lfs reset` or `git lfs stash`, adding new sentences to explain existing commands seem preferable to me.

- My issue with it, is whether `git lfs migrate` best describe what users are doing. I was not importing large files to pointers per se, I was simply telling git to discard local changes and pull remote as-is. (They just happened to be large files and pointers.)

- Also, is "*discarding local changes*" a side-effect of `--skip-fetch`, or its main effect?

- [The manual for `--skip-fetch`](https://github.com/git-lfs/git-lfs/blob/master/docs/man/git-lfs-migrate.1.ronn) is too technical for a layman user like me to understand, which I suspect caused its low adoption. (I have searched issues and googled it, very few mentions `--skip-fetch`.)

**Comment by bk2204** [comment]
The goal of `--skip-fetch` is simply to avoid an expensive fetch operation.  It's not strictly required, since you would specify only `HEAD` and therefore wouldn't modify any other paths, but fetching may not work when disconnected from the Internet and it doesn't do anything useful in this case.

**Comment by ttaylorr** [comment]
> the best way to handle this might be a `git lfs disable` command

I like that idea. It seems like a good solution to: "Git LFS is trying very hard to do what it thinks is the right thing" and "just let me take over for a second." `git lfs uninstall && git reset --hard HEAD` works, too, but the trouble is that the user has to then remember to run `git lfs install` again. I think _that's_ where the user experience is sub-par.

I imagine that `git lfs disable` would work in the same way (though I'm not clear on what `git lfs disable reset --hard` would do, other than run the arguments as a Git command, in which I might recommend calling the command `git lfs unsafe -- <program run without Git LFS>`). The benefit is that we can leave all of the Git LFS filters/hooks in place, and write our state to `.git/config`, in say, `lfs.disabled = 1`.

Leaving the hooks in place, if we try and push with `lfs.disabled = 1`, we can prevent the push from going forward from the `pre-push` command, by first checking our state to make sure that we're not disabled. In this case, we can provide some guidance to the user, or the user can bypass us again with `git push --no-verify`.

> Maybe something like `git lfs restore-pointers <rev>`? That can take any dirty files that should be pointers and convert them back.

I think that this is a reasonable idea, but it seems that it might be more special-purpose than we typically implement in a Git LFS command. In other words, it seems to provide a subset of the functionality that a user might want to execute with `git lfs disable`. That said, I'm not sure what the _rest_ of that set looks like :-). So perhaps we can take the best of both worlds: implement `git lfs disable`, and give the user a shortcut like: `git lfs disable --restore-pointers`, which executes:

```ShellSession
$ git lfs disable
$ git reset --hard HEAD
$ git lfs enable
```

and then returns control to the user.

> No matter what, I think it would be worthwhile to look into adding some info on this and how to fix it to an error message somewhere.

I agree, but I think a good approach here would be to make that command `man git-lfs-migrate` (or `git lfs migrate --help`). Given the destructive nature of `git lfs migrate import [...]`, I am concerned that a user encountering this scenario will see a command surrounded in backticks, and immediately run it without being fully informed of what it will do to their repository.

Perhaps this is all an argument to improve the interaction when actually running `git lfs migrate` (e.g., "Are you sure you want to ...?", and so on), but I agree at the very least we need (1) better visibility on its existence, and (2) some sort of control passing mechanism for Git LFS.

**Comment by RiweiChen** [comment]
1、comment out the .gitattributes file
```
#*.a filter=lfs diff=lfs merge=lfs -text
#*.lib filter=lfs diff=lfs merge=lfs -text
#*.wav filter=lfs diff=lfs merge=lfs -text
#*.dat filter=lfs diff=lfs merge=lfs -text
#*.model filter=lfs diff=lfs merge=lfs -text
#*.caffemodel filter=lfs diff=lfs merge=lfs -text
#*.zip filter=lfs diff=lfs merge=lfs -text
#*.bin filter=lfs diff=lfs merge=lfs -text
```
2、checkout .gitattributes file
```
git checkout .gitattributes 
```
3、the error is disappear
Actually, I do nothing at all.

**Comment by saschpe** [comment]
> 1、comment out the .gitattributes file
> 2、checkout .gitattributes file
> 3、the error is disappear
> Actually, I do nothing at all.

This was the only way to fix it for me. All other approaches discussed didn't work. As the issue creator mentioned, this often happens when switching to (feature) branches that are behind the integration branch and when some LFS-tracked files have been changed meanwhile.
